libsolidity/ast/Types.cpp

```diff
@@ -2143,47 +2142,84 @@ TypeResult StructType::interfaceType(bool _inLibrary) const
 	if (!_inLibrary && m_interfaceType.is_initialized())
 		return *m_interfaceType;
 
-	TypePointer result = TypePointer{};
+	bool recursion = false;
+	TypeResult result{TypePointer{}};
 
-	if (_inLibrary && location() == DataLocation::Storage)
-		result = shared_from_this();
-	else if (recursive())
-		result = TypePointer{};
-	else
+	auto visitor = [&](
+		StructDefinition const& _struct,
+		CycleDetector<StructDefinition>& _cycleDetector,
+		size_t /*_depth*/
+	)
 	{
 		// Check that all members have interface types.
-		// We pass "false" to canBeUsedExternally (_inLibrary), because this struct will be
-		// passed by value and thus the encoding does not differ, but it will disallow
-		// mappings.
-		// Also return false if at least one struct member does not have a type.
+		// Return false if at least one struct member does not have a type.
 		// This might happen, for example, if the type of the member does not exist,
 		// which is reported as an error.
-		bool allOkay = true;
-		for (auto const& var: m_struct.members())
+		for (ASTPointer<VariableDeclaration> const& variable: _struct.members())
 		{
 			// If the struct member does not have a type return false.
 			// A TypeError is expected in this case.
-			if (!var->annotation().type)
+			if (!variable->annotation().type)
 			{
-				allOkay = false;
-				break;
+				result = TypeResult::err("Invalid type!");
+				return;
 			}
-			if (!var->annotation().type->interfaceType(false).get())
+
+			Type const* memberType = variable->annotation().type.get();
+
+			while (dynamic_cast<ArrayType const*>(memberType))
+				memberType = dynamic_cast<ArrayType const*>(memberType)->baseType().get();
+
+			if (StructType const* innerStruct = dynamic_cast<StructType const*>(memberType))
+				if (_cycleDetector.run(innerStruct->structDefinition()))
+				{
+					recursion = true;
+					if (_inLibrary && location() == DataLocation::Storage)
+						continue;
+					else
+					{
+						result = TypeResult::err("Recursive struct are required to have storage data location.");
```

Comment:
"Recursive structs can only be passed as storage pointers to libraries, not as memory objects to contract functions.'